### PR TITLE
__cuda_array_interface__

### PR DESCRIPTION
**Pull request by sklam**
for interop with 3rd party CUDA array.

### CODE REVIEW

**Comment by seibert** [comment]
I can confirm this works!
``` python
import cupy

from numba import cuda

@cuda.jit
def add(x, y, out):
    start = cuda.grid(1)
    stride = cuda.gridsize(1)
    for i in range(start, x.shape[0], stride):
        out[i] = x[i] + y[i]

a = cupy.arange(10)
b = a * 2
out = cupy.empty_like(a)

print('out before:', out)

add[1, 32](a, b, out)

print('out after:', out)
print('array types:', type(a), type(b), type(out))
```

Output:
```
(cupy_dev) seibert@nvidia1:~/repos/numba_cupy$ python demo.py
out before: [0 0 0 0 0 0 0 0 0 0]
out after: [ 0  3  6  9 12 15 18 21 24 27]
array types: <class 'cupy.core.core.ndarray'> <class 'cupy.core.core.ndarray'> <class 'cupy.core.core.ndarray'>
```

**Comment by seibert** [comment]
It looks like support for `@vectorize`/`@guvectorize` needs to be done still.

**Comment by seibert** [comment]
I think you forgot to check in `cuda_array_interface.rst`  :)

**Comment by sklam** [comment]
Oops, added missing file now

File: docs/source/cuda/memory.rst [hunk 1]
```diff
@@ -20,6 +20,17 @@ transfer:
 .. autofunction:: numba.cuda.to_device
    :noindex:
 
+In addition to the device arrays, numba can consume any object that implements
+:ref:`cuda array interface <cuda-array-interface>`.  These objects can be
+converted into a device array by creating a view of the GPU buffer using the
```
Comment (seibert) [inline docs/source/cuda/memory.rst]: I think we should phrase this as "These objects also can be manually converted into a Numba device array by creating a view of the GPU buffer using the following APIs:"

File: docs/source/cuda/memory.rst [hunk 2]
```diff
@@ -20,6 +20,17 @@ transfer:
 .. autofunction:: numba.cuda.to_device
    :noindex:
 
+In addition to the device arrays, numba can consume any object that implements
```
Comment (seibert) [inline docs/source/cuda/memory.rst]: s/numba/Numba/

**Comment by seibert** [comment]
Two minor documentation nits, and then this looks good to merge.

**Pull request merged**